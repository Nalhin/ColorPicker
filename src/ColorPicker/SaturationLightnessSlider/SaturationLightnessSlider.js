import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import ColorSelector from '../SliderHandle';
import { SELECTED_SLIDER, SELECTOR_SIZE } from '../constants';
import { checkBoundaries, colorArrayToRgba } from '../Utils';

const Background = styled.div`
  height: 200px;
  display: block;
  background-image: linear-gradient(0deg, rgba(0, 0, 0, 1), rgba(0, 0, 0, 0)),
    linear-gradient(90deg, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0));
`;

const SaturationLightnessSlider = ({
  hueColor,
  selected,
  changeSelected,
  changeSatLightPosition,
  satLightRef,
  satLightColor,
  satLightPosition,
}) => {
  const setPosition = React.useCallback(
    event => {
      const satLight = satLightRef.current;
      const rect = satLight.getBoundingClientRect();
      const newX = Math.floor(
        (event.clientX !== undefined
          ? event.clientX
          : event.touches[0].clientX) - rect.left,
      );
      const newY = Math.floor(
        (event.clientY !== undefined
          ? event.clientY
          : event.touches[0].clientY) - rect.top,
      );
      const SCOPE = Object.freeze({
        X_MIN: 0,
        X_MAX: satLight.offsetWidth,
        Y_MIN: 0,
        Y_MAX: satLight.offsetHeight,
      });
      const [x, y] = checkBoundaries(newX, newY, SCOPE);
      changeSatLightPosition({ x, y: y - SELECTOR_SIZE / 2 });
    },
    [changeSatLightPosition, satLightRef],
  );

  const getPositionMove = React.useCallback(
    event => {
      event.preventDefault();
      if (selected === SELECTED_SLIDER.SATURATION_LIGHTNESS) {
        setPosition(event);
      }
    },
    [selected, setPosition],
  );

  const getPositionClick = event => {
    event.preventDefault();
    setPosition(event);
    changeSelected(SELECTED_SLIDER.SATURATION_LIGHTNESS);
  };

  React.useEffect(() => {
    if (selected === SELECTED_SLIDER.SATURATION_LIGHTNESS)
      window.addEventListener('mousemove', getPositionMove);
    else window.removeEventListener('mousemove', getPositionMove);
    return () => {
      window.removeEventListener('mousemove', getPositionMove);
    };
  }, [getPositionMove, selected]);

  return (
    <div>
      <ColorSelector
        color={satLightColor}
        position={satLightPosition}
        click={getPositionClick}
      >
        <Background
          ref={satLightRef}
          style={{
            backgroundColor: colorArrayToRgba(hueColor),
          }}
        />
      </ColorSelector>
    </div>
  );
};

SaturationLightnessSlider.propTypes = {
  hueColor: PropTypes.arrayOf(PropTypes.number).isRequired,
  satLightColor: PropTypes.arrayOf(PropTypes.number).isRequired,
  selected: PropTypes.number.isRequired,
  changeSelected: PropTypes.func.isRequired,
  satLightRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
  ]).isRequired,
  changeSatLightPosition: PropTypes.func.isRequired,
  satLightPosition: PropTypes.shape({
    x: PropTypes.number,
    y: PropTypes.number,
  }).isRequired,
};

export default SaturationLightnessSlider;
