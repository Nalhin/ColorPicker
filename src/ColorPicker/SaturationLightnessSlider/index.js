import SaturationLightnessSlider from './SaturationLightnessSlider';

export default SaturationLightnessSlider;
