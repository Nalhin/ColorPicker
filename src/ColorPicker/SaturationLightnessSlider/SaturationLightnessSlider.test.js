import React from 'react';
import { render } from '@testing-library/react';
import SaturationLightnessSlider from './SaturationLightnessSlider';

describe('SaturationLightnessSlider', () => {
  const props = {
    hueColor: [255, 240, 220, 1],
    selected: 0,
    changeSelected: () => {},
    changeSatLightPosition: () => {},
    satLightRef: () => {},
    satLightColor: [255, 240, 220, 1],
    satLightPosition: { x: 40, y: 50 },
  };

  it('Should render correctly', () => {
    const { container } = render(<SaturationLightnessSlider {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
