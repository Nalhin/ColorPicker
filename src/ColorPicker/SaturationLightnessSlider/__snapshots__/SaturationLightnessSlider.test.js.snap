// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaturationLightnessSlider Should render correctly 1`] = `
<div>
  <div
    class="css-cqd1um-SliderHandle__Wrapper e9y9zbe0"
  >
    <div
      class="css-gjeaqo-SliderHandle__Handle e9y9zbe1"
      style="top: 50px; left: 33px; background: rgb(255, 240, 220);"
    />
    <div
      class="css-xhn5x9-SaturationLightnessSlider__Background eq818vc0"
      style="background-color: rgb(255, 240, 220);"
    />
  </div>
</div>
`;
